compiler/src/dotty/tools/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -702,33 +706,18 @@ trait BCodeBodyBuilder extends BCodeSkelBuilder {
             if (t.symbol ne defn.Object_synchronized) genTypeApply(t)
             else genSynchronized(app, expectedType)
 
-        case Apply(fun @ DesugaredSelect(Super(_, _), _), args) =>
-          def initModule(): Unit = {
-            // we initialize the MODULE$ field immediately after the super ctor
-            if (!isModuleInitialized &&
-              jMethodName == INSTANCE_CONSTRUCTOR_NAME &&
-              fun.symbol.javaSimpleName == INSTANCE_CONSTRUCTOR_NAME &&
-              claszSymbol.isStaticModuleClass) {
-              isModuleInitialized = true
-              mnode.visitVarInsn(asm.Opcodes.ALOAD, 0)
-              mnode.visitFieldInsn(
-                asm.Opcodes.PUTSTATIC,
-                thisName,
-                str.MODULE_INSTANCE_FIELD,
-                "L" + thisName + ";"
-              )
-            }
-          }
+        case Apply(fun @ DesugaredSelect(Super(superQual, _), _), args) =>
           // 'super' call: Note: since constructors are supposed to
           // return an instance of what they construct, we have to take
           // special care. On JVM they are 'void', and Scala forbids (syntactically)
           // to call super constructors explicitly and/or use their 'returned' value.
           // therefore, we can ignore this fact, and generate code that leaves nothing
           // on the stack (contrary to what the type in the AST says).
-          mnode.visitVarInsn(asm.Opcodes.ALOAD, 0)
+
+          // scala/bug#10290: qual can be `this.$outer()` (not just `this`), so we call genLoad (not just ALOAD_0)
+          genLoad(superQual)
```

Comment:
Since you are doing this for the JVM back-end, could you also apply the appropriate fix in the SJS back-end?

The code to fix is at
https://github.com/lampepfl/dotty/blob/6ada79cfe3f736b3d00c7f47882d4e023a053976/compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala#L1391-L1401
to be updated similarly to what was done for scalac at
https://github.com/scala-js/scala-js/commit/e02799b636b577eb56593d08c21a7cb2e68719e8#diff-c9a195a739d51194e79f0655bc93bd61R2343-R2363
and you can enable the `RegressionTest.scala` test (i.e., remove it from the blacklist) to test the fix at
https://github.com/lampepfl/dotty/blob/6ada79cfe3f736b3d00c7f47882d4e023a053976/project/Build.scala#L1024

Reply:
Thanks for the detailed instruction @sjrd, I'll do that.

Reply:
@sjrd FYI, in this PR (unlike scalac), the `qual` in a `Super` is still `this`, it's only in the backend phase where the invariant is broken. Do you think the changes proposed above are still relevant?

Reply:
Actually we can enrich the `-Ycheck` to make sure it's always a compiler invariant. Do you think it's a good idea @smarter ?

Reply:
Sounds good to me.

Reply:
> @sjrd FYI, in this PR (unlike scalac), the `qual` in a `Super` is still `this`, it's only in the backend phase where the invariant is broken. Do you think the changes proposed above are still relevant?

The proposed changes are direct equivalent of the changes you made to the JVM back-end, but in terms of SJSIR. So if the changes you made are relevant for the JVM (which I assume is true since you did them), then the proposed changes for the JS back-end are also relevant.